### PR TITLE
chore(sealed_blocks): optimize fetching of sealed block header at a given height

### PR DESCRIPTION
**Pull request by rymnc**
## Linked Issues/PRs

- #2023

## Description

We don't need to fetch the header *and* the consensus header from the storage if either one of them is `None`. If `consensus` evaluates to `None`, we can early return a `None`, but if the consensus header exists, we can then fetch the block header.

Performs another optimization in instances where we fetch the `SealedBlock` and only end up using the `entity` wrapped within it, and not the consensus headers. Previously, both the consensus headers and the block were fetched from the db and then returned as a `SealedBlock`, Now, we just check for existence of a given block height in the `SealedBlockConsensus` table, and avoid the allocation for the consensus headers. 

> note: also restricts how many blocks worth of transactions can be requested over p2p

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests
- [x] [The specification](https://github.com/FuelLabs/fuel-specs/) matches the implemented behavior (link update PR if changes are needed)

### Before requesting review
- [x] I have reviewed the code myself
- [x] I have created follow-up issues caused by this PR and linked them here

### After merging, notify other teams

[Add or remove entries as needed]

- [ ] [Rust SDK](https://github.com/FuelLabs/fuels-rs/)
- [ ] [Sway compiler](https://github.com/FuelLabs/sway/)
- [ ] [Platform documentation](https://github.com/FuelLabs/devrel-requests/issues/new?assignees=&labels=new+request&projects=&template=NEW-REQUEST.yml&title=%5BRequest%5D%3A+) (for out-of-organization contributors, the person merging the PR will do this)
- [ ] Someone else?

### CODE REVIEW

**Comment by rymnc** [comment]
added block range restriction in 0a7b23af6 since the CI runners are struggling with all the open PRs

**Comment by rymnc** [comment]
p2p metrics issue: https://github.com/FuelLabs/fuel-core/issues/808

File: crates/fuel-core/src/database/sealed_block.rs [hunk 1]
```diff
@@ -51,6 +54,23 @@ impl OnChainIterableKeyValueView {
         }
     }
 
+    /// Returns a `Block`, which is sealed, by height
+    pub fn get_block_by_height(
+        &self,
+        height: &BlockHeight,
+    ) -> StorageResult<Option<Block>> {
+        if self
```
Comment (xgreenx) [inline crates/fuel-core/src/database/sealed_block.rs:62]: Why do we want to do an additional lookup into `SealedBlockConsensus` table? We can go directly to the `self.get_full_block`

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Changed
 - [2106](https://github.com/FuelLabs/fuel-core/pull/2106): Remove deadline clock in POA and replace with tokio time functions.
+- [2112](https://github.com/FuelLabs/fuel-core/pull/2112): Alter the way the sealed blocks are fetched with a given height.
```
Comment (xgreenx) [inline CHANGELOG.md:16]: We need to move it to `## [Unreleased]` section

File: bin/fuel-core/src/cli/run/p2p.rs [hunk 3]
```diff
@@ -66,6 +66,10 @@ pub struct P2PArgs {
     #[clap(long = "max-headers-per-request", default_value = "100", env)]
     pub max_headers_per_request: u32,
 
+    /// Max number of blocks in a transaction request response
+    #[clap(long = "max-blocks-per-tx-request", default_value = "100", env)]
+    pub max_blocks_per_tx_request: u32,
```
Comment (xgreenx) [inline bin/fuel-core/src/cli/run/p2p.rs:71]: I think we can reuse `max_headers_per_request` to reduce number of arguments

File: crates/fuel-core/src/database/sealed_block.rs [hunk 4]
```diff
@@ -51,6 +54,23 @@ impl OnChainIterableKeyValueView {
         }
     }
 
+    /// Returns a `Block`, which is sealed, by height
```
Comment (netrome) [inline crates/fuel-core/src/database/sealed_block.rs:57]: I find this docstring a bit confusing as we don't return a `SealedBlock` in this function. We chatted offline about this, and the important thing here is that the block has been sealed (i.e. it is signed) but downstream we only need the actual block.

```suggestion
    /// Returns a `Block` by it's height. Only returns a block if it has been sealed.
```

Comment (netrome) [reply crates/fuel-core/src/database/sealed_block.rs:57]: New information based on @xgreenx's comment below and an offline discussion. The block importer will only persist blocks that are sealed so we can omit this from the docstring altogether
```suggestion
    /// Returns a `Block` by its height.
```

File: crates/fuel-core/src/database/sealed_block.rs (hunk 1 above)
Comment (rymnc) [reply crates/fuel-core/src/database/sealed_block.rs]: addressed in 5498c5fd9

File: bin/fuel-core/src/cli/run/p2p.rs (hunk 3 above)
Comment (rymnc) [reply bin/fuel-core/src/cli/run/p2p.rs]: addressed in https://github.com/FuelLabs/fuel-core/pull/2112/commits/06262b9b8c210ac8b485d96050f52a54d96d18f4

File: CHANGELOG.md (hunk 2 above)
Comment (rymnc) [reply CHANGELOG.md]: addressed in https://github.com/FuelLabs/fuel-core/pull/2112/commits/da650f4ebe72bb42003d859750eb84cbf26a2271

File: crates/services/p2p/src/service.rs [hunk 5]
```diff
@@ -437,6 +437,20 @@ where
...
+                if range.len() > max_len {
+                    tracing::error!("Requested range of blocks is too big. Requested length: {:?}, Max length: {:?}", range.len(), max_len);
+                    let response = None;
+                    let _ = self.p2p_service.send_response_msg(
+                        request_id,
+                        ResponseMessage::Transactions(response),
+                    );
+                    return Ok(())
+                }
+
```
Comment (netrome) [inline crates/services/p2p/src/service.rs:453]: Nit: This is a very long match arm. While it was pretty long already before this PR, I'd find the code easier to read if we could break this out to a separate function. If we use `tracing::instrument` on that function we could have a cleaner error message, where the lengths are automatically added to the span.

**Review by MitchTurner — APPROVED** [review]

File: crates/services/p2p/src/service.rs (hunk 5 above)
Comment (rymnc) [reply crates/services/p2p/src/service.rs:453]: https://github.com/FuelLabs/fuel-core/issues/2121

Comment (netrome) [reply crates/services/p2p/src/service.rs:453]: Thanks for opening!

**Review by netrome — APPROVED** [review]
Just a few nits and enhancement suggestions from me, otherwise the PR looks good!

File: crates/services/p2p/src/service.rs [hunk 6]
```diff
@@ -437,6 +437,20 @@ where
         let response_channel = self.request_sender.clone();
         match request_message {
             RequestMessage::Transactions(range) => {
+                let max_len = self
+                    .max_blocks_per_request
+                    .try_into()
+                    .expect("u32 should always fit into usize");
+                if range.len() > max_len {
+                    tracing::error!("Requested range of blocks is too big. Requested length: {:?}, Max length: {:?}", range.len(), max_len);
```
[the remark this answers is not in the document]
Comment (rymnc) [reply crates/services/p2p/src/service.rs]: addressed in ebca20dd5, with the other usage of this as well

**Review by MitchTurner — APPROVED** [review]

**Review by MitchTurner — APPROVED** [review]

File: crates/services/p2p/src/service.rs [hunk 7]
```diff
@@ -437,6 +437,24 @@ where
         let response_channel = self.request_sender.clone();
         match request_message {
             RequestMessage::Transactions(range) => {
+                let max_len = self
```
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:440]: We can use `u32` instead of `usize` to avoid `try_into`

**Review by xgreenx — APPROVED** [review]

**Pull request merged**